Dodanie domeny ArticleFile wraz z logiką uploadu na MinIO

adam-szalanski opened the pull request.

Commit:
feat: podwaliny pod service connection dla minIO

adam-szalanski committed:
feat: dodanie domeny dla ArticleFile oraz uploadu pliku na MinIO

adam-szalanski committed:
refactor: drobne poprawki

adam-szalanski committed:
refactor: poprawki wgrywania pliku

adam-szalanski committed:
refactor: dodanie beana tworzącego topic kafka

adam-szalanski committed:
test: testy publikowania na kafkę oraz wysyłania pliku na minio

adam-szalanski committed:
test: testy repo postgresa

jacchm commented on src/main/java/org/aszjch/demoapp/aplication/ArticleDto.java:
```diff
@@ -1,11 +1,17 @@
-  @Null(message = "id must not be provided.")
-  private Long id;
-
+    @Null(message = "id must not be provided.")
```
Zamiast walidacji na null, może lepiej dać @JsonIgnore? wtedy na pewno to nie wpadnie z JSON'a, a skoro to zawsze ma być null to może nie ma potrzeby tego tak zostawiać?

jacchm commented on src/main/java/org/aszjch/demoapp/aplication/MapperHelper.java:
```diff
@@ -0,0 +1,32 @@
+    private final TempFilesService tempFilesService;
+
+    @Named("getCurrentTime")
+    OffsetDateTime getCurrentTime(Object ignore) {
```
Czy ten Object ignore jest na potrzeby mapstructa?

adam-szalanski replied:
Yup, mapstruct jest trochę upośledzony pod tym względem i wymaga przy mapowaniu, żeby wepchnąć do metody jakieś source. Szukałem wcześniej po necie, to piszą, że się nie da obejść dziadostwa, przynajmniej jeżeli chcę użyć dependency injection opartego o Named/QualifiedByName

jacchm commented on src/main/java/org/aszjch/demoapp/aplication/ArticleFileController.java:
```diff
@@ -0,0 +1,37 @@
+    @PostMapping
+    ResponseEntity<String> uploadFile(@RequestParam Long articleId, @RequestPart MultipartFile uploadedFile) {
+        log.info("Creating articleFile");
+        ArticleFile articleFile = mapper.toEntity(articleId, uploadedFile);
```
Jestem przyzwyczajony do dawania final wszędzie w parametrach i lokalnych zmiennych. Wtedy jak faktycznie jakąś zmieniasz w metodzie (np. bo masz pętle) to lepiej to widać. Zmień sobie ustawienia IntelliJ żeby Ci wstawiało finale :)

jacchm commented on src/main/java/org/aszjch/demoapp/aplication/ArticleFileDtoMapper.java:
```diff
@@ -0,0 +1,22 @@
+import static org.mapstruct.InjectionStrategy.CONSTRUCTOR;
+
+@Mapper(componentModel = "spring", injectionStrategy = CONSTRUCTOR, uses = MapperHelper.class)
+interface ArticleFileDtoMapper {
```
ArticleFileMapper

jacchm commented on src/main/java/org/aszjch/demoapp/aplication/ArticleFileDtoMapper.java:
```diff
@@ -0,0 +1,22 @@
+
+import static org.mapstruct.InjectionStrategy.CONSTRUCTOR;
+
+@Mapper(componentModel = "spring", injectionStrategy = CONSTRUCTOR, uses = MapperHelper.class)
```
Możemy wynieść to co można do MapstructConfig

jacchm commented on src/main/java/org/aszjch/demoapp/domain/article/ArticleServiceImpl.java:
```diff
@@ -0,0 +1,108 @@
+
+    @Override
+    public Article update(Long id, Article article) {
+        Article entity = getById(id).orElse(new Article());
```
Chcemy robić UPSERT? Czy klasyczek - PUT jako full update tylko?
IMHO możemy zostawić :)

adam-szalanski replied:
Wcześniej rozmawialiśmy o upsert, brzmiał bardzo sensownie, więc czemu nie 😄

adam-szalanski replied:
Niby istnieje możliwość, że ktoś mógłby przywrócić w taki sposób artykuł, który w międzyczasie usunął, ale w kontekście tego projektu raczej nie robi to dużej różnicy 🤔

jacchm commented on src/main/java/org/aszjch/demoapp/domain/article/ArticleServiceImpl.java:
```diff
@@ -0,0 +1,108 @@
+    @Override
+    public void assignFile(ArticleFile articleFile) {
+        Optional<Article> oEntity = getById(articleFile.getArticleId());
+        if (oEntity.isPresent()) {
```
Panie Panie. oEntity nic mi nie mówi jako nazwa zmiennej. Chyba taką standardową praktyką jest dodawanie postfixu "Opt" więc byłoby articleOpt.
Druga sprawa: optionale zostały zrobione nie po to, żeby walić ify w każdym miejscu, chociaż to wygodne. Pobawmy się trochę Javą :) i zrób to ładnie w streamie Optional'a - .ifPresent() metoda albo .ifPresentOrElse()

jacchm commented on src/main/java/org/aszjch/demoapp/domain/article/ArticleServiceImpl.java:
```diff
@@ -0,0 +1,108 @@
+    @Override
+    public void deleteFile(Long id) {
+        Optional<Article> oEntity = getById(id);
+        if (oEntity.isPresent()) {
```
j.w.

jacchm commented on src/main/java/org/aszjch/demoapp/domain/article/ArticleServiceImpl.java:
```diff
@@ -0,0 +1,108 @@
+    public Article create(Article article) {
+        Article saved = repository.save(article);
+        publisher.publish(
+                String.valueOf(new ArticleCreatedDto(saved.getId(), saved.getTitle(), saved.getAuthor(),
```
Nie o to do końca mi chodziło mówiąc: "weź zamień tego Stringa na obiekt". Chodziło mi o to, żeby metoda publish faktycznie obiekt przyjmowała :D

jacchm commented on src/main/java/org/aszjch/demoapp/domain/article/ArticleServiceImpl.java:
```diff
@@ -0,0 +1,108 @@
+    @AllArgsConstructor
+    @Getter
+    @Setter
+    private static class ArticleCreatedDto implements Serializable {
```
Raczej przeniesiemy to do osobnej klasy w ogóle, będziemy z tego korzystać na zewnątrz.

jacchm commented on src/main/java/org/aszjch/demoapp/domain/article/port/ArticlePublisher.java:
```diff
@@ -0,0 +1,7 @@
+
+public interface ArticlePublisher {
+
+    void publish(String message);
```
publish(NaszObiekt obiekt)

jacchm commented on src/main/java/org/aszjch/demoapp/domain/articlefile/port/ArticleFilePublisher.java:
```diff
@@ -0,0 +1,8 @@
+
+import org.aszjch.demoapp.domain.articlefile.ArticleFile;
+
+public interface ArticleFilePublisher {
```
Nie podoba mi się w nazwie Publisher tutaj. ArticleFileStorage brzmi IMHO lepiej. Pomyśl, czy się zgadzasz, może masz lepszą nazwę :)

adam-szalanski replied:
ArticleFileStorageService?

jacchm commented on src/main/java/org/aszjch/demoapp/infrastructure/config/GeneralBeanConfig.java:
```diff
@@ -0,0 +1,15 @@
+import java.time.Clock;
+
+@Configuration
+class GeneralBeanConfig {
```
Zmieniłbym nazwę na ClockConfig. Niech nazwa mówi co się dzieje :)

jacchm commented on src/main/java/org/aszjch/demoapp/infrastructure/config/KafkaTopicsConfig.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.kafka.config.TopicBuilder;
+
+@Configuration
+class KafkaTopicsConfig {
```
KafkaConfig po prostu

jacchm commented on src/main/java/org/aszjch/demoapp/infrastructure/config/MinIOAutoConfiguration.java:
```diff
@@ -0,0 +1,29 @@
+@AutoConfiguration
+@EnableConfigurationProperties(MinIOProperties.class)
+@Slf4j
+public class MinIOAutoConfiguration {
```
MinIOConfig prawdopodobnie wystarczy

adam-szalanski replied:
Pewnie tak, ale rejestruję to w resources jako AutoConfiguration obok innych takich ze springa. Każdy z nich ma nazwę z ujętym AutoConfiguration, więc tutaj chciałbym tą końcówkę jednak zostawić 😉

jacchm commented on src/main/java/org/aszjch/demoapp/infrastructure/config/TempFilesProperties.java:
```diff
@@ -0,0 +1,16 @@
+@Configuration
+@ConfigurationProperties(prefix = "temp")
+@Getter
+@Setter
```
Na pewno chcemy tutaj setter mieć?

adam-szalanski replied:
Bez settera chyba nie zadziała, jakbym chciał ustawić inną ścieżkę przez propertiesy. Popraw mnie proszę, jeżeli się mylę

adam-szalanski replied:
Wyjebałem ten setter ostatecznie

jacchm commented on src/main/java/org/aszjch/demoapp/infrastructure/storage/ArticleFileMinIOPublisherImpl.java:
```diff
@@ -0,0 +1,84 @@
+        externalStorageService.upload(multipartFile);
+    }
+
+    public static class CustomMultipartFile implements MultipartFile {
```
Chyba to mamy w osobnej klasie?

jacchm commented on src/main/java/org/aszjch/demoapp/infrastructure/storage/CustomMultipartFile.java:
```diff
@@ -0,0 +1,62 @@
+import java.io.InputStream;
+import java.nio.file.Files;
+
+public class CustomMultipartFile implements MultipartFile {
```
Chyba to mamy wyżej w klasie wew.

adam-szalanski replied:
Rzeczywiście, jakoś magicznie musiałem je zduplikować... Już wyrzucam jedno z nich

jacchm commented on src/main/java/org/aszjch/demoapp/infrastructure/storage/ExternalStorageService.java:
```diff
@@ -0,0 +1,9 @@
+import io.minio.ObjectWriteResponse;
+import org.springframework.web.multipart.MultipartFile;
+
+public interface ExternalStorageService {
```
Kurde, nie podoba mi się serwajs wszędzie. Spróbujmy to lepiej ponazywać. Wydaje mi się, że tutaj ExternalStorage wystarczy, a klasa implementująca to np. MinIOStorage.

adam-szalanski replied:
Może ExternalStorageClient? Jak ktoś jest upośledzony jak ja, to widząc ExternalStorage pomyśli, że to już jest ten storage i pliki znajdują się wewnątrz klasy, a nie, że jest to klient służący do komunikacji z tym storage'm

jacchm commented on src/main/java/org/aszjch/demoapp/infrastructure/storage/TempFilesService.java:
```diff
@@ -0,0 +1,43 @@
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class TempFilesService {
```
Czy to Serwajs?

adam-szalanski replied:
W sumie to nie, jakaś pokraka z tego wyszła trochę. Niby converter by pasował patrząc po nazwie pliku, ale znowu tak nie do końca, skoro on przegrywa treść z multiparta na zwykły file...

jacchm commented on src/test/java/org/aszjch/demoapp/TestClockBean.java:
```diff
@@ -0,0 +1,19 @@
+public class TestClockBean {
+
+    @Bean
+    @Primary
```
Najs ;)

jacchm commented on src/test/java/org/aszjch/demoapp/TestcontainersConfiguration.java:
```diff
@@ -1,32 +1,42 @@
+    }
+
+    @Bean
+    @ServiceConnection
```
Super że to zakodziłeś, sztos

jacchm commented on src/test/java/org/aszjch/demoapp/domain/article/port/KafkaArticlePublisherIT.java:
```diff
@@ -0,0 +1,38 @@
+    private ArticlePublisher kafkaArticlePublisher;
+    @SpyBean
+    private KafkaTestListener kafkaTestListener;
+    @Captor
```
Hmmm ciekawe użycie Captor'a. Być może jest to sensowne podejście :) zobaczymy jak się sprawdzi.

adam-szalanski replied:
Właśnie tak sobie badam trochę te captory ostatnio i np. jak chcę sprawdzić wysyłkę po rest na jakiś system, to stawiam sobie controller w testach i łapię argument captorem to co do niego przyjdzie. Wtedy mogę sobie łatwo zobaczyć, czy json na 200% poprawnie się sformatował, czy tak jak tutaj, mogę sprawdzić, czy po kafce przyszła wiadomość poprawnie.

Najbardziej do mnie przemawia w takim podejściu to, że nie muszę pchać dodatkowych technologii, jak jakieś RestAssured, czy inny MockWebServer.

jacchm commented on src/test/java/org/aszjch/demoapp/domain/article/port/KafkaTestListener.java:
```diff
@@ -0,0 +1,15 @@
+@Slf4j
+class KafkaTestListener {
+
+    @KafkaListener(topics = {"articles"})
```
lepsza nazwa topika konieczna - articles-created np.
Nie musimy całej konwencji wymyślać, ale niech to będzie coś sensownego.

jacchm commented on src/test/java/org/aszjch/demoapp/domain/articlefile/port/MinIOArticleFilePublisherIT.java:
```diff
@@ -0,0 +1,63 @@
+    private MinIOConnetionDetails minIOConnetionDetails;
+
+    @Test
+    void publishes_article_file_to_minio(@TempDir File tempDir) throws Exception {
```
ta konwencja nazewnicza jest z dupy. Możemy camel casem pisać, zamiast pieprzyć się z podłogami.

jacchm commented on src/test/java/org/aszjch/demoapp/domain/articlefile/port/MinIOArticleFilePublisherIT.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Test
+    void publishes_article_file_to_minio(@TempDir File tempDir) throws Exception {
+        String fileName = "test-file.txt";
```
Złota zasada testów - jak najkrótsze. Jako zasadę "na oko" przyjmij sobie, że jak test ma więcej niż 7-10 linijek to coś jest nie tak. Tworzenie obiektów warto wynosić do prywatnych choćby metod. A jak tworzysz ich dużo na wskroś testów to tworzyć klasy dla nich.

jacchm commented on src/test/java/org/aszjch/demoapp/domain/articlefile/port/MinIOArticleFilePublisherIT.java:
```diff
@@ -0,0 +1,63 @@
+        minIOArticleFilePublisher.publish(articleFile);
+
+
+        Iterable<Result<Item>> results = minioClient.listObjects(ListObjectsArgs.builder()
```
To też da się skrócić do 1 asercji na pewno, bez tego całego szumu wokoło.

adam-szalanski replied:
Zrobiłem skróconą wersję opartą o assertDoesNotThrow i taki check jak tutaj: https://stackoverflow.com/a/72626422

jacchm commented on src/test/java/org/aszjch/demoapp/domain/article/port/PostgresArticleRepositoryTest.java:
```diff
@@ -0,0 +1,141 @@
+
+    @Test
+    void saves_article() {
+        Article article = new Article();
```
Zrób metodę prywatną na tworzenie artykułów.

jacchm commented on src/test/java/org/aszjch/demoapp/domain/article/port/PostgresArticleRepositoryTest.java:
```diff
@@ -0,0 +1,141 @@
+
+    @Test
+    void finds_article_by_id() {
+        Article article = new Article();
```
Ten artykuł powinien być wczytany już na BD. Nie powinno się w testach korzystać z metod które też testujemy. Powinniśmy testcontainerki wykorzystać do wgrania odpowiednich data setów testowych.
To co tutaj robimy jest antypatternem, lepiej się tego pozbyć.

jacchm commented on src/main/java/org/aszjch/demoapp/domain/article/Article.java:
```diff
@@ -0,0 +1,19 @@
+import java.time.OffsetDateTime;
+
+
+// TODO: important - DOMAIN CLASS STATE CAN BE CHANGED ONLY FROM WITHIN THE CLASS, NOT FROM OUTSIDE
```
todo do wywalenia.

jacchm commented:
Chyba mamy konflikty z powodu poprzedniego MRa. możliwe że nie odbiłeś się dobrze branchem, albo 2 raz te zmiany chcą wejść.

adam-szalanski committed:
refactor(ArticleDto): wyrzucenie walidacji na nulla

adam-szalanski committed:
refactor(KafkaTopicConfig): zmiana nazwy klasy na KafkaConfig

adam-szalanski committed:
refactor: zmiana nazwy topica kafka na articles-created

adam-szalanski committed:
refactor: wstępne poprawki po review

adam-szalanski committed:
fix: poprawka konfliktów po rebase

adam-szalanski committed:
test(MinIOArticleFileStorageServiceIT): skrócenie testu publikacji pliku na minio

adam-szalanski committed:
doc(Article): wywalenie todo

adam-szalanski committed:
refactor: poprawka testów postgresa

adam-szalanski committed:
refactor: zmiana konwencji nazewniczej testów na cammelCase

adam-szalanski committed:
refactor: przeniesienie configu kafka do yamla w main + naprawa kafki

Z kafką był ten problem, o którym wspominałem, czyli że trzeba dodać package z obiektem DTO do trusted packages + zmiana serializerów/deserializerów na jsonowe ze springa

jacchm commented on pom.xml:
```diff
@@ -83,6 +83,11 @@
+        <dependency>
+            <groupId>io.minio</groupId>
+            <artifactId>minio</artifactId>
+            <version>8.5.11</version>
```
do zmiennej, wyrób sobie nawyk wrzucania do zmiennych

adam-szalanski replied:
Poprawka, do wyjebania, jakieś zduplikowane dependency po rozwiązywaniu konfliktów zostało

jacchm commented on src/main/java/org/aszjch/demoapp/aplication/ArticleController.java:
```diff
@@ -45,10 +52,10 @@ ResponseEntity<IdDTO> createArticle(@RequestBody final ArticleDto articleDto) {
+        final Article updated = service.update(id, article);
 
         return ResponseEntity
                 .ok()
```
tutaj widzę mamy błąd, pewnie mój - PUT to noContent()

adam-szalanski replied:
O i widzisz, tego akurat nie wiedziałem, że PUT też noContent powinien mieć. Dobrze wiedzieć, bo sugerowałem się upsertem, że może powinien id zwracać

jacchm commented on src/main/java/org/aszjch/demoapp/aplication/ArticleDto.java:
```diff
@@ -1,11 +1,17 @@
 
+@Getter
+@Setter
 class ArticleDto {
```
Nie wiem jaka jest polityka taka super poprawna, ale wydaje mi się, że wsadzanie @Data z automatu jest okej. Masz wtedy od razu @ToString i @EqualsAndHashcode prócz tego co teraz. A one lubią się przydawać :) 
Ja z reguły walę @Data chociaż możemy też tak zostawić.

adam-szalanski replied:
Ja nie jestem zbyt dużym fanem pakowania @Data, bo zawsze się sonar pluje, że data w entity to zły pomysł, bo wydłuży działanie niektórych metod. Tutaj entity nie jest, problemu żadnego z tego nie będzie, więc wprowadzę data

jacchm commented on src/main/java/org/aszjch/demoapp/aplication/ArticleDtoMapper.java:
```diff
@@ -1,16 +1,17 @@
     @Mapping(ignore = true, target = "id")
-    Article toEntity(ArticleDto articleDto);
+    @Mapping(ignore = true, target = "filename")
+    Article toEntity(final ArticleDto articleDto);
```
Jeśli używamy tego przy PUTcie też, to ten filename ignore=true nam wyrzuci cenną informację.

adam-szalanski replied:
Jest używany przy PUT, ale tylko po to, żeby przerobić DTO przychodzące przez controller na Article. Na tym etapie nigdy nie ma filename w DTO (ono nawet nie ma tego pola), więc nic nie wyrzuci

jacchm commented on src/main/java/org/aszjch/demoapp/aplication/ArticleFileController.java:
```diff
@@ -0,0 +1,38 @@
+@Slf4j
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/articles/files")
```
Ja nie lubię takiego budowania ścieżek, wiem ze robimy tak też w robocie, ale IMHO to jest syfiate i niezgodne z restem. Jak dla mnie powinno być /article-files. Jak kiedyś znajdę to podeślę Ci dobre praktyki resta.

jacchm replied:
Generalnie taka forma /articles/files mówi, że operujesz na kolekcji files z artykułów, ale wszystkich (?) bo nie ma /{id}/ w środku. jest to turbo dziwna konstrukcja z perspektywy budowania REST API

jacchm commented on src/main/java/org/aszjch/demoapp/aplication/ArticleFileMapper.java:
```diff
@@ -0,0 +1,20 @@
+
+    @Mapping(target = "file", source = "uploadedFile", qualifiedByName = {"MapperHelper", "multipartFileToFile"})
+    @Mapping(target = "filename", source = "uploadedFile", qualifiedByName = "getFileName")
+    ArticleFile toEntity(final Long articleId, final MultipartFile uploadedFile);
```
Ciekawy case mi wpadł do głowy - jak usuniemy artykuł to powinniśmy też od razu usuwać article file - mamy to zrealizowane :D ?

adam-szalanski replied:
Nah, nie mamy, głównie dlatego, że nie mamy też usuwania artykułu xD
Idę dostrugać

adam-szalanski replied:
Dobra, przy pierwszym podejściu się zajebałem o zapętlenie dependency injection. Musze rozjebać funkcjonalność ArticleFileAssignmentService osobno od tego ArticleServiceImpl jednak

jacchm commented on src/main/java/org/aszjch/demoapp/aplication/ArticleFileMapper.java:
```diff
@@ -0,0 +1,20 @@
+    @Mapping(target = "filename", source = "uploadedFile", qualifiedByName = "getFileName")
+    ArticleFile toEntity(final Long articleId, final MultipartFile uploadedFile);
+
+    @Named("getFileName")
```
Najs, ładnie śmigasz z tymi @Named i ogólnie z mapstructem :)

jacchm commented on src/main/java/org/aszjch/demoapp/aplication/IdDTO.java:
```diff
@@ -1,7 +1,8 @@
-  }
+public record IdDTO(Long id) {
+
+    static IdDTO of(final Long id) {
```
Najs Adam! nie wiem czy wiesz - ale to jest statyczna metoda konstrukcyjna :) taki pattern. String.valueOf() jest przykładem takiej. Java ją wykorzystuje w wielu miejscach.

adam-szalanski replied:
Dzięki dzięki, tak jak pisałem na instagramie, spod Twojej łapki ta klasa wyszła xD

jacchm commented on src/main/java/org/aszjch/demoapp/aplication/MapperHelper.java:
```diff
@@ -0,0 +1,32 @@
+@RequiredArgsConstructor
+@Named("MapperHelper")
+@Component
+class MapperHelper {
```
To jest malutka klasa, ale generalnie skoro się bawimy to chciałbym, żebyśmy bardziej przechylili szale w kierunku podejścia dogmatycznego, żeby sobie uświadomić/wyrobić pewne nawyki. 
wstrzykiwanie tutaj zegara i tego konwertera robi nam 2 odpowiedzialności klasy. Mówiłem Ci kiedyś o parametrze spójności klasy - czyli czy wszystkie pola są wykorzystywane w każdej metodzie - tutaj ewidentnie jest podział. Nie lubię też szczerze mówiąc nazwy Helper - powinniśmy raczej z tego nie korzystać, bo co to k**** jest helper tak na prawdę? :) Nic ta nazwa nie mówi. Wyodrębnij te dwie odpowiedzialności do osobnych klas z ładnymi nazwami. Da się to zrobić 100%

jacchm commented on src/main/java/org/aszjch/demoapp/domain/article/ArticleCreatedDto.java:
```diff
@@ -0,0 +1,32 @@
+    private String abstractText;
+
+    @Override
+    public String toString() {
```
Co tutaj robi toString a na górze milion adnotacji? Można wrzucić @Data - polecam zawsze poczytać doku lomboka, ona ładnie opisuje co się dzieje :)

jacchm commented on src/main/java/org/aszjch/demoapp/domain/article/ArticleServiceImpl.java:
```diff
@@ -0,0 +1,73 @@
+
+@Slf4j
+@RequiredArgsConstructor
+public class ArticleServiceImpl implements ArticleService, ArticleFileAssignmentService {
```
public wylatuje sprzed klasy. Interfejs musi być publiczny, implementacja już niekoniecznie (metody tak ale klasa nie).

jacchm commented on src/main/java/org/aszjch/demoapp/domain/article/ArticleServiceImpl.java:
```diff
@@ -0,0 +1,73 @@
+    private final ArticlePublisher publisher;
+
+    @Override
+    public List<Article> get() {
```
Zapnijmy tutaj tez Paginację i Sortowanie żeby cokolwiek się działo. Filtrowanie olejmy - zobaczymy jak wyjdzie nam przepychanie obiektów - pewnie trzeba będzie zmapować :)

adam-szalanski replied:
Na pewno chcemy pchać Page ze springframework? Domena chyba miała być w miarę wolna od takich rzeczy

adam-szalanski replied:
Tego Page to na razie odpuszczam, żeby więcej czasu na przypadki testowe poświęcić

jacchm commented on src/main/java/org/aszjch/demoapp/domain/article/ArticleServiceImpl.java:
```diff
@@ -0,0 +1,73 @@
+    @Override
+    public Article update(final Long id, final Article article) {
+        final Article entity = getById(id).orElse(new Article());
+        entity.setId(id);
```
setId jest tutaj okej, ale ta reszta wygląda jak jakieś mapowanie trochę - wrzuć do nowej metody klasy Article z parametrem Article :) przypisywanie wszystkiego poniżej. Plus findById() powinno zwracać optionala i powinniśmy być w stanie ładnie to w Optionalu obsłużyć wszystko :)

jacchm replied:
Możliwe że będzie można użyć mapstructa żeby nadpisał tylko istniejące w obiekcie wejściowym pola - wtedy połączymy oba obiekty ładnie :)

jacchm commented on src/main/java/org/aszjch/demoapp/domain/article/ArticleServiceImpl.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Override
+    public void assignFile(final ArticleFile articleFile) {
```
Nie wiem czy nie chcemy jakiejś zwrotki z tego. To się powinno zachowywać tak samo jak update tylko trochę taki specyficzny. Zastanowimy się nad tym wspólnie jak już będą testy i reszta. Dodaj gdzieś do listy tematów do dyskusji.

jacchm commented on src/main/java/org/aszjch/demoapp/domain/articlefile/ArticleFileServiceImpl.java:
```diff
@@ -0,0 +1,20 @@
+    @Override
+    public void upload(final ArticleFile articleFile) {
+        articleFileStorageService.store(articleFile);
+        articleFileAssignmentService.assignFile(articleFile);
```
Jesteś pewien że coś co uploaduje powinno przypisywać plik? a nie powinien to robić wywołujący ten serwis u siebie?

adam-szalanski replied:
To jest bardziej klasa sterująca tak naprawdę. Jej by się przydało lepszą nazwę znaleźć. Próbowałem coś się do niej dobrać, ale zakurwiłem głową w circular dependency i uznałem, że zacznę od poprawienia reszty. Jutro pomyślę co da się z tym zrobić

jacchm commented on src/main/java/org/aszjch/demoapp/infrastructure/config/ArticleFileConfig.java:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class ArticleFileConfig {
```
Te klasy config nie są potrzebne public chyba - powywalaj na maksa public stad i innych Config plików

jacchm commented on src/main/java/org/aszjch/demoapp/infrastructure/config/TempFilesProperties.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@ConfigurationProperties(prefix = "temp")
```
Po co tutaj te @ConfigurationPropeties skoro mamy private final który bierze z System.getProperty?

adam-szalanski replied:
Chciałem robić konfigurowalne, jakbyśmy chcieli gdzieś indziej trzymać pliki tymczasowe. No ale chuj, raczej nie będziemy chcieli z tego co widzę

jacchm commented on src/main/java/org/aszjch/demoapp/infrastructure/storage/CustomMultipartFile.java:
```diff
@@ -0,0 +1,62 @@
+
+    private final File file;
+
+    public CustomMultipartFile(final File file) {
```
na pewno public constructor? pamiętaj ze konstruktor to nie normalna metoda

jacchm commented on src/main/java/org/aszjch/demoapp/infrastructure/storage/MultipartFileToFileConverter.java:
```diff
@@ -0,0 +1,43 @@
+        }
+    }
+
+    public File convertMultipartFile(final MultipartFile multipartFile) throws IOException {
```
Zastanawiam się czy tych bajerów nie da się upchnąć w naszej klasie CustomMultiPartFile ale na razie oglądam to na githubie to ciężko powiedzieć.

adam-szalanski replied:
Wszystko zależy jaki MultipartFile nam w łapki wpadnie. Jak ten nasz customowy, to tak naprawdę wystarczyłby nam getFile na nim. Natomiast jeżeli wpadnie faktyczny MultipartFile springowy, to już trochę ciężko będzie...

jacchm commented on src/test/java/org/aszjch/demoapp/domain/article/port/PostgresArticleRepositoryTest.java:
```diff
@@ -0,0 +1,100 @@
+
+@SpringBootTest
+@Import({TestcontainersConfiguration.class, TestClockBean.class})
+class PostgresArticleRepositoryTest {
```
Testy obczaję później :)

jacchm commented on src/main/java/org/aszjch/demoapp/domain/article/ArticleServiceImpl.java:
```diff
@@ -0,0 +1,73 @@
+            repository.save(article);
+            log.info("File [{}] has been successfully assigned to article [{}].", articleFile.getFilename(),
+                     article.getId());
+        }, () -> log.error("File assignment failed. Article id {} not found", articleFile.getArticleId()));
```
has not been found.

jacchm commented on src/main/java/org/aszjch/demoapp/domain/article/ArticleServiceImpl.java:
```diff
@@ -0,0 +1,73 @@
+        getById(id).ifPresentOrElse(article -> {
+            article.setFilename(null);
+            repository.save(article);
+        }, () -> log.error("Article id {} not found", id));
```
has not been found.

jacchm commented on src/main/java/org/aszjch/demoapp/domain/article/ArticleServiceImpl.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Override
+    public void deleteFile(final Long id) {
```
Raczej unassignFile?

adam-szalanski committed:
refactor: wywalenie duplikatu dependency minio

adam-szalanski committed:
refactor: zmiana zwrotki w put na no content

adam-szalanski committed:
refactor: data w ArticleDto

adam-szalanski committed:
refactor: poprawka ścieżki dla ArticleFileController

adam-szalanski committed:
refactor: dodanie usuwania pliku z minio

adam-szalanski committed:
refactor: rozdzielenie MapperHelpera oraz uprzątnięcie ArticleCreatedDto

adam-szalanski committed:
refactor: wywalenie public z ArticleFileServiceImpl

adam-szalanski committed:
refactor: upsert robiony mapstructem

adam-szalanski committed:
refactor: poprawki widoczności oraz wyrzucenie @ConfigurationProperties

adam-szalanski committed:
refactor: wyrzucenie public z constructor

adam-szalanski committed:
refactor: has not been found

adam-szalanski committed:
refactor: unassign file

adam-szalanski committed:
feat: usuwanie pliku z minio oraz odpinanie go od Article w bazie

adam-szalanski commented:
Odbijam branch pod przykłady testowe od tego momentu
